File: SolaceApp/Services/FirestoreResourceService.swift
import Foundation
import FirebaseFirestore
import SolaceCore

/// Firebase-backed implementation of `ResourceServicing`. Articles are
/// editorial content stored in Firestore (seeded via the console); relaxation
/// exercises are self-contained content shipped in the app, since they don't
/// need remote updates the way articles do.
final class FirestoreResourceService: ResourceServicing {
    private let firestore = Firestore.firestore()

    func fetchResources() async throws -> [ResourceItem] {
        let snapshot = try await firestore.collection("resources").getDocuments()
        return snapshot.documents.compactMap { document in
            let data = document.data()
            guard
                let title = data["title"] as? String,
                let summary = data["summary"] as? String,
                let body = data["body"] as? String,
                let category = ResourceCategory(rawValue: data["category"] as? String ?? "")
            else { return nil }
            return ResourceItem(
                id: document.documentID,
                title: title,
                summary: summary,
                body: body,
                category: category,
                tags: data["tags"] as? [String] ?? []
            )
        }
    }

    func fetchRelaxationExercises() async throws -> [RelaxationExercise] {
        RelaxationExercise.breathingExercises + RelaxationExercise.groundingExercises
    }
}

File: SolaceAndroid/app/src/main/java/com/jonathanalumasa/solace/service/firebase/FirestoreResourceService.kt
package com.jonathanalumasa.solace.service.firebase

import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.jonathanalumasa.solace.model.ResourceCategory
import com.jonathanalumasa.solace.model.ResourceItem
import com.jonathanalumasa.solace.service.ResourceService
import kotlinx.coroutines.tasks.await

/**
 * Firebase-backed [ResourceService]. The `resources` collection is curated
 * content, public-read and seeded server-side.
 */
class FirestoreResourceService(
    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance()
) : ResourceService {

    override suspend fun fetchResources(): List<ResourceItem> =
        firestore.collection("resources")
            .get()
            .await()
            .documents
            .mapNotNull(::resource)

    private fun resource(document: DocumentSnapshot): ResourceItem? {
        val title = document.getString("title") ?: return null
        @Suppress("UNCHECKED_CAST")
        val tags = document.get("tags") as? List<String> ?: emptyList()
        return ResourceItem(
            id = document.id,
            title = title,
            summary = document.getString("summary") ?: "",
            body = document.getString("body") ?: "",
            category = ResourceCategory.fromRaw(document.getString("category"))
                ?: ResourceCategory.GENERAL,
            tags = tags
        )
    }
}
